File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/Python/src/rlglue/environment/EnvironmentLoader.py
# 
# Copyright (C) 2007, Mark Lee
# 
#http://rl-glue-ext.googlecode.com/
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#
#  $Revision: 473 $
#  $Date: 2009-01-29 22:50:12 -0500 (Thu, 29 Jan 2009) $
#  $Author: brian@tannerpages.com $
#  $HeadURL: http://rl-glue-ext.googlecode.com/svn/trunk/projects/codecs/Python/src/rlglue/environment/EnvironmentLoader.py $

import sys
import os
import rlglue.network.Network as Network
from ClientEnvironment import ClientEnvironment


from rlglue.versions import get_svn_codec_version
from rlglue.versions import get_codec_version

def loadEnvironment(theEnvironment):
	theSVNVersion=get_svn_codec_version()
	theCodecVersion=get_codec_version()
	client = ClientEnvironment(theEnvironment)

	host = Network.kLocalHost
	port = Network.kDefaultPort

	hostString = os.getenv("RLGLUE_HOST")
	portString = os.getenv("RLGLUE_PORT")

	if (hostString != None):
		host = hostString

	try:
		port = int(portString)
	except TypeError:
		port = Network.kDefaultPort

	print "RL-Glue Python Environment Codec Version: "+theCodecVersion+" (Build "+theSVNVersion+")"
	print "\tConnecting to " + host + " on port " + str(port) + "..."
	sys.stdout.flush()

	client.connect(host, port, Network.kRetryTimeout)
	print "\t Environment Codec Connected"

	client.runEnvironmentEventLoop()
	client.close()

def loadEnvironmentLikeScript():
#Assumes you've already done the checking that the number of args and such is good
	envModule = __import__(sys.argv[1])
	envClass = getattr(envModule,sys.argv[1])
	env = envClass()

	loadEnvironment(env)

File: Domains/InfiniteMario/15-rl-competition-2009/system/codecs/Java/src/org/rlcommunity/rlglue/codec/util/EnvironmentLoader.java
/* 
 * Copyright (C) 2007, Brian Tanner
 * 
http://rl-glue-ext.googlecode.com/
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
 * 
 *  $Revision: 802 $
 *  $Date: 2009-02-27 21:42:44 -0500 (Fri, 27 Feb 2009) $
 *  $Author: brian@tannerpages.com $
 *  $HeadURL: http://rl-glue-ext.googlecode.com/svn/trunk/projects/codecs/Java/src/org/rlcommunity/rlglue/codec/util/EnvironmentLoader.java $
 * 
 */
package org.rlcommunity.rlglue.codec.util;

import java.net.InetAddress;
import org.rlcommunity.rlglue.codec.EnvironmentInterface;
import org.rlcommunity.rlglue.codec.RLGlueCore;
import org.rlcommunity.rlglue.codec.network.ClientEnvironment;
import org.rlcommunity.rlglue.codec.network.Network;

/**
 * This class can be called from the command line to load an environment and create
 * an executable RL environment program.  
 * 
 * We've recently refactored it to make it useful if anyone ever wants to create
 * local instances of network-bound environment from inside a JVM (like Matlab)
 * @author btanner
 */
public class EnvironmentLoader implements Runnable {

    String host = Network.kDefaultHost;
    int port = Network.kDefaultPort;
    final EnvironmentInterface theEnvironment;
    ClientEnvironment theClient = null;

    public EnvironmentLoader(EnvironmentInterface theEnvironment) {
        assert theEnvironment != null : "theEnvironment is null in EnvironmentLoader Constructor";
        this.theEnvironment = theEnvironment;
        String envVariableHostString = System.getenv("RLGLUE_HOST");
        String envVariablePortString = System.getenv("RLGLUE_PORT");
        setHostAndPorts(envVariableHostString, envVariablePortString);
    }

    public EnvironmentLoader(String hostString, String portString, EnvironmentInterface theEnvironment) {
        this(theEnvironment);
        setHostAndPorts(hostString, portString);
    }

    /**
     * Try these new settings.  We'll only actually set them if they seem valid.
     * @param hostString
     * @param portString
     */
    private void setHostAndPorts(String hostString, String portString) {

        //Now override the default or env variable port and string with these specific settings
        if (hostString != null) {
            try {
                InetAddress theAddress = InetAddress.getByName(hostString);
                host = hostString;
            } catch (Exception e) {
                System.err.println("Problem resolving requested hostname: " + hostString + " so using default.");
            }
        }

        if (portString != null) {

            try {
                int parsedPort = Integer.parseInt(portString);

                if (parsedPort < 0 || parsedPort > 65535) {
                    System.err.println("Could not use port you requested: " + parsedPort + " is not a valid port number.\n");
                } else {
                    port = parsedPort;
                }
            } catch (Exception e) {
                System.err.println("Could not use port you requested: " + portString + " could not be parsed as an int.");
                e.printStackTrace();
            }
        }
    }

    public void killProcess() {
        theClient.killProcess();
    }

    public void run() {
        String ImplementationVersion = RLGlueCore.getImplementationVersion();
        String SpecVersion = RLGlueCore.getSpecVersion();

        System.out.println("RL-Glue Java Environment Codec Version: " + SpecVersion + " (" + ImplementationVersion + ")");
        System.out.println("\tConnecting to " + host + " on port " + port + "...");

        theClient = new ClientEnvironment(theEnvironment);
        try {
            theClient.connect(host, port, Network.kRetryTimeout);
            System.out.println("\tEnvironment Codec Connected");
            theClient.runEnvironmentEventLoop();
            theClient.close();
        } catch (Exception e) {
            System.err.println("EnvironmentLoader run(" + theEnvironment.getClass() + ") threw Exception: " + e);
            e.printStackTrace();
        }
    }

    /**
     * Loads the class envClassName as an rl-glue environment.
     * @param envClassName
     */
    public static EnvironmentLoader loadEnvironment(String envClassName) {
        EnvironmentInterface env = null;

        //Had to use the system classloader in case the codec was actually installed.
        //If its installed, it uses a different classloader.
        try {
            env = (EnvironmentInterface) ClassLoader.getSystemClassLoader().loadClass(envClassName).newInstance();
        } catch (Exception e) {
            System.err.println("loadEnvironment(" + envClassName + ") threw Exception: " + e);
            e.printStackTrace();
            System.exit(1);
        }

        return new EnvironmentLoader(env);
    }

    public static void main(String[] args) throws Exception {
        String usage = "java EnvironmentLoader <Environment> -classpath <Path To RLGlue>";

        String envVars = "The following environment variables are used by the environment to control its function:\n" +
                "RLGLUE_HOST  : If set the environment will use this ip or hostname to connect to rather than " + Network.kDefaultHost + "\n" +
                "RLGLUE_PORT  : If set the environment will use this port to connect on rather than " + Network.kDefaultPort + "\n";

        if (args.length < 1) {
            System.out.println(usage);
            System.out.println(envVars);
           System.exit(1);
        }
        EnvironmentLoader theEnvironmentLoader = loadEnvironment(args[0]);
        theEnvironmentLoader.run();
    }
}
